src/workerd/api/node/buffer.c++

```diff
@@ -0,0 +1,396 @@
+  } else {
+    offset = kj::min(buffer.size(), offset);
+  }
+
+  const auto indexOfImpl = [&](kj::ArrayPtr<kj::byte> needle) -> kj::Maybe<uint32_t> {
+    if (needle.size() == 0) {
+      return offset;
+    }
+
+    if (buffer.size() == 0) {
+      return nullptr;
+    }
+
+    if ((needle.size() + offset) > buffer.size()) {
+      return nullptr;
+    }
+
+    // TODO(perf): We could be more sophisticated here and use a more efficient
+    // search algorithm rather than using the default. For now, this is good
+    // enough and we can revisit later.
+    auto it = !findLast ?
+        std::search(buffer.begin() + offset, buffer.end(), needle.begin(), needle.end()) :
+        std::find_end(buffer.begin() + offset, buffer.end(), needle.begin(), needle.end());
+    if (it == buffer.end()) return nullptr;
+    return it - buffer.begin();
+  };
+
+  KJ_SWITCH_ONEOF(value) {
+    KJ_CASE_ONEOF(string, v8::Local<v8::String>) {
+      auto decoded = decodeString(js, string, kj::mv(encoding).orDefault(kj::str("utf8"_kj)));
+      return indexOfImpl(decoded);
+    }
+    KJ_CASE_ONEOF(source, jsg::BufferSource) {
+      return indexOfImpl(source.asArrayPtr());
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+void BufferUtil::swap(jsg::Lock& js, kj::Array<kj::byte> buffer, int size) {
+  if (buffer.size() <= 1) return;
+  switch (size) {
+    case 16: return SwapBytes<uint16_t>(buffer);
+    case 32: return SwapBytes<uint32_t>(buffer);
+    case 64: return SwapBytes<uint64_t>(buffer);
+  }
+  KJ_UNREACHABLE;
+}
+
+v8::Local<v8::String> BufferUtil::toString(
+    jsg::Lock& js,
+    kj::Array<kj::byte> bytes,
+    uint32_t start,
+    uint32_t end,
+    kj::String encoding) {
+  auto slice = bytes.slice(start, end);
+  if (slice.size() == 0) return v8::String::Empty(js.v8Isolate);
+  switch (getEncoding(encoding)) {
+    case Encoding::ASCII:
+      // While ASCII is not Latin1, Node.js treats these as somewhat equivalent cases
+      // so we just fall through...
+    case Encoding::LATIN1: {
+      return jsg::v8StrFromLatin1(js.v8Isolate, slice);
+    }
+    case Encoding::UTF8: {
+      return jsg::v8Str(js.v8Isolate, slice.asChars());
+    }
+    case Encoding::UTF16LE: {
+      // If the buffer length is not an even multiple of 2, the last byte
+      // will be ignored and omitted from the results.
+      uint16_t* view = reinterpret_cast<uint16_t*>(slice.begin());
+      return jsg::v8Str(js.v8Isolate, kj::ArrayPtr<uint16_t>(view, slice.size() / 2));
+    }
+    case Encoding::BASE64: {
+      return jsg::v8Str(js.v8Isolate, kj::encodeBase64(slice));
+    }
+    case Encoding::BASE64URL: {
+      return jsg::v8Str(js.v8Isolate, kj::encodeBase64Url(slice));
+    }
+    case Encoding::HEX: {
+      return jsg::v8Str(js.v8Isolate, kj::encodeHex(slice));
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+uint32_t BufferUtil::write(
+    jsg::Lock& js,
+    kj::Array<kj::byte> buffer,
+    v8::Local<v8::String> string,
+    uint32_t offset,
```

Comment:
What would happen if I passed `-1` as an argument? Will it actually be `0xffff` in this case or something more dangerous/correct?

Reply:
The JS/TS side handles basic input validation here, ensuring that `offset` is not negative. However, the `amountToCopy` calculation just below this should handle also just in case:

```
auto amountToCopy = kj::min(kj::min(length, decoded.size()), buffer.size() - offset);
```

Reply:
To be extra safe, we can change the `KJ_DASSERT` into `KJ_ASSERT` so that checking is not limited to debug builds.